File: frontend/src/data/notifications.ts
// Cola de notificaciones (datos reales: router /notifications).
//
// fetchNotifications (cola activa + conteo de no-leídas) + mark/dismiss/read-all para la página
// /notificaciones y el AlertBell. `toAlertEvent` proyecta un aviso persistido al shape de la campana
// (la campana unifica alertas dinámicas + avisos persistidos). El deep-link lo decide el emisor
// (campo del contrato), no un mapa por kind acá.

import { apiGet, apiPost } from "@/lib/api"
import type { AlertEvent, PersistedNotification } from "@/types/domain"

interface NotificationApi {
  id: number
  kind: string
  severity: PersistedNotification["severity"]
  title: string
  body: string
  payload: Record<string, unknown>
  deep_link: string | null
  created_at: string
  read_at: string | null
  dismissed_at: string | null
  expires_at: string | null
}

interface NotificationListApi {
  items: NotificationApi[]
  unread: number
  next_cursor: number | null
}

function toNotification(n: NotificationApi): PersistedNotification {
  return {
    id: n.id,
    kind: n.kind,
    severity: n.severity,
    title: n.title,
    body: n.body,
    payload: n.payload,
    deepLink: n.deep_link,
    createdAt: n.created_at,
    readAt: n.read_at,
    dismissedAt: n.dismissed_at,
    expiresAt: n.expires_at,
  }
}

export interface NotificationFeed {
  items: PersistedNotification[]
  unread: number
  nextCursor: number | null
}

/** Cola activa de avisos del usuario (newest-first) + conteo de no-leídas para el badge. */
export async function fetchNotifications(): Promise<NotificationFeed> {
  const page = await apiGet<NotificationListApi>("/notifications?limit=100")
  return {
    items: page.items.map(toNotification),
    unread: page.unread,
    nextCursor: page.next_cursor,
  }
}

/** Marca un aviso como leído (sale del conteo de no-leídas). */
export async function markNotificationRead(id: number): Promise<void> {
  await apiPost(`/notifications/${id}/read`)
}

/** Descarta un aviso (lo saca de la cola activa). */
export async function dismissNotification(id: number): Promise<void> {
  await apiPost(`/notifications/${id}/dismiss`)
}

/** Marca leídos todos los avisos activos sin leer. */
export async function readAllNotifications(): Promise<void> {
  await apiPost("/notifications/read-all")
}

/** Proyecta un aviso persistido a un item del AlertBell. El id se prefija con `notif:` para no
 *  colisionar con los ids de las alertas dinámicas y para que el provider rutee read/dismiss. */
export function toAlertEvent(n: PersistedNotification): AlertEvent {
  return {
    id: `notif:${n.id}`,
    severity: n.severity,
    kind: n.kind,
    title: n.title,
    detail: n.body,
    at: n.createdAt,
    read: n.readAt !== null,
    deepLink: n.deepLink ?? "/notificaciones",
    source: "persisted",
    notifId: n.id,
  }
}

File: src/memex/api/routers/notifications.py
"""Lectura y gestión de la cola de notificaciones (`GET /notifications` + acciones).

Superficie de LECTURA del servicio de notificaciones; la escritura (encolar) la hace el
`PersistentNotifier` por el seam `Notifier`, no por HTTP. Alimenta la página `/notificaciones` y el
AlertBell del dashboard. Multi-tenant: todo se acota a `current_user_id`.
"""

from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query

from memex.api.auth import current_user_id
from memex.api.schemas import NotificationList
from memex.db import connection
from memex.logging import get_logger
from memex.notifications import store

router = APIRouter(prefix="/notifications", tags=["notifications"])

UserID = Annotated[int, Depends(current_user_id)]

_log = get_logger("memex.api.notifications")


@router.get("", response_model=NotificationList)
async def list_notifications(
    user_id: UserID,
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    cursor: int | None = Query(default=None, description="id < cursor (paginación newest-first)"),
) -> dict[str, Any]:
    """Cola activa de avisos del usuario (newest-first). Excluye descartados y vencidos; incluye el
    conteo de no-leídas para el badge de la campana. Paginación por cursor (`id < :cursor`)."""
    with connection() as conn:
        items = store.list_active(conn, user_id=user_id, limit=limit, cursor=cursor)
        unread = store.count_unread(conn, user_id=user_id)
    next_cursor = items[-1]["id"] if len(items) == limit else None
    return {"items": items, "unread": unread, "next_cursor": next_cursor}


@router.post("/read-all")
async def mark_all_notifications_read(user_id: UserID) -> dict[str, int]:
    """Marca leídos todos los avisos activos sin leer del usuario. Devuelve cuántos cambió."""
    with connection() as conn:
        updated = store.mark_all_read(conn, user_id=user_id)
    _log.info("notifications.read_all", user_id=user_id, updated=updated)
    return {"updated": updated}


@router.post("/{notification_id}/read")
async def mark_notification_read(notification_id: int, user_id: UserID) -> dict[str, bool]:
    """Marca un aviso como leído (sale del conteo de no-leídas). 404 si no es del usuario."""
    with connection() as conn:
        ok = store.mark_read(conn, notification_id=notification_id, user_id=user_id)
    if not ok:
        raise HTTPException(status_code=404, detail="notification not found")
    return {"ok": True}


@router.post("/{notification_id}/dismiss")
async def dismiss_notification(notification_id: int, user_id: UserID) -> dict[str, bool]:
    """Descarta un aviso (lo saca de la cola activa). 404 si no es del usuario."""
    with connection() as conn:
        ok = store.dismiss(conn, notification_id=notification_id, user_id=user_id)
    if not ok:
        raise HTTPException(status_code=404, detail="notification not found")
    return {"ok": True}
